### ISSUE TITLE
iRule CRD and allow attaching iRule to Virtual Server Resources

### ISSUE THREAD

**Issue by mitchellmaler** [F5Networks/k8s-bigip-ctlr#1571]
#### Title
iRule CRD and allow attaching iRule to Virtual Servers

#### Description
The ability to create iRules from Kubernetes Custom Resources and attaching those to Virtual Servers.

#### Actual Problem
We have some specific routing logic that the ability to manage these at the k8s layer along with allowing tenants to write their own using custom resources would make it a smooth process.

#### Solution Proposed
Custom Resources for iRules and allow attaching those to Virtual Server Custom Resources.

**Comment by mdditt2000** [comment]
@mitchellmaler Good morning. You looking for something like the following?

```
apiVersion: "cis.f5.com/v1"
kind: CRD
metadata:
  name: my-crd
spec:
  virtualServerAddress: "172.16.3.4"
  iRules:
    - /Common/test-irule.  --------------------- add here!
  selector:
    matchLabels:
      app: nginx
```
If this is what you want i can get this added. Please let me know.

**Comment by mitchellmaler** [comment]
Yes being able to add irules to Virtual Servers would be perfect. 

Another idea is also the ability to author irules from kubernetes using a new custom resource definition like below. This would allow tenants to write their own rules and then attach to virtual server resources all from k8s.

```
apiVersion: "cis.f5.com/v1"
kind: iRule
metadata:
  name: my-new-irule
  namespace: my-namespace
spec:
  ruleDefinition: |
    when HTTP_REQUEST {
        if { [HTTP::uri] starts_with "/abc/" } {
            HTTP::uri [string map {"/abc/" "/xyz/"} [HTTP::uri]]
        }
    }

```

**Comment by mdditt2000** [comment]
@mitchellmaler nice. Would you prefer the CRD to create the iRule as shown in your example or simply reference the iRule form BIG-IP? Referencing the iRule is way easier but that object would need to be on BIG-IP or CIS would receive a 422 error

**Comment by mitchellmaler** [comment]
I think both options should be available. At least for starters/short-term the ability to reference existing iRules on the Big-IP would be great. Then if in the future or part of this request the ability to author brand new irules using a kubernetes custom resource and then reference those would allow tenants to create and use their own without the need to add it to the bigip outside kubernetes before creating the virtual server.

**Comment by mdditt2000** [comment]
@mitchellmaler created internal Jira for PM tracking [CRD] iRule reference for CRD CONTCNTR-2252. Sprint planning is next week. Would like to get into next release.

**Comment by mdditt2000** [comment]
@mitchellmaler on a side note are you using GTM/DNS with LTM?

**Comment by mdditt2000** [comment]
iRule support for TransportServer CRD https://github.com/F5Networks/k8s-bigip-ctlr/issues/1570

**Comment by mitchellmaler** [comment]
@mdditt2000 Today we are currently only using Big-IP LTM pairs for our environments.

**Comment by mitchellmaler** [comment]
@mdditt2000 Is there an update on adding iRule Reference to VS and a new iRule CRD support?

**Comment by mdditt2000** [comment]
@mitchellmaler flagged for CIS 2.3.

**Comment by trinaths** [comment]
This issue is fixed with CIS v2.3
